[Go] out of bounds panics in RefResolver and MetaStringResolver

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/fory/issues) and found no similar issues.


### Version

0.17.0

### Component(s)

Go

### Minimal reproduce step

Run these tests- (just place them inside a test file and run test script)
```go
func TestRefResolver_OOBPanic(t *testing.T) {
	resolver := newRefResolver(true) // Enable tracking
	buffer := NewByteBuffer(nil)
	
	// Craft a buffer with RefFlag (-2) followed by a huge RefID (9999)
	buffer.WriteInt8(RefFlag)
	buffer.WriteVarUint32(9999)
	buffer.SetReaderIndex(0)

	var ctxErr Error
	// This currently PANICS. The fix should prevent the panic.
	require.NotPanics(t, func() {
		resolver.ReadRefOrNull(buffer, &ctxErr)
	}, "RefResolver.GetReadObject should not panic on OOB index")
}
func TestMetaStringResolver_NegativeIndexPanic(t *testing.T) {
	resolver := NewMetaStringResolver()
	buffer := NewByteBuffer(nil)
	
	// header = 1 means (header & 1 != 0) is true (it's a reference)
	// and length = header >> 1 = 0.
	// index = length - 1 = -1.
	buffer.WriteVarUint32Small7(1)
	buffer.SetReaderIndex(0)

	var ctxErr Error
	// This currently PANICS. The fix should prevent the panic.
	require.NotPanics(t, func() {
		_, err := resolver.ReadMetaStringBytes(buffer, &ctxErr)
		if err == nil {
			// If it didn't panic, it should at least return an error
			t.Errorf("Expected error for negative index, got nil")
		}
	}, "MetaStringResolver should not panic on negative index")
}

```

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

@chaokunyang that's the failing error -
```bash
❯ go test -v -count=1 -run "TestRefResolver_OOBPanic|TestMetaStringResolver_NegativeIndexPanic" 2>&1 | tail -40
                                        /usr/local/go/src/testing/testing.go:2101 +0x4c5
                Test:           TestRefResolver_OOBPanic
                Messages:       RefResolver.GetReadObject should not panic on OOB index
--- FAIL: TestRefResolver_OOBPanic (0.00s)
=== RUN   TestMetaStringResolver_NegativeIndexPanic
    security_test.go:57: 
                Error Trace:    security_test.go:57
                Error:          func (assert.PanicTestFunc)(0x65e920) should not panic
                                        Panic value:    runtime error: index out of range [-1]
                                        Panic stack:    goroutine 8 [running]:
                                runtime/debug.Stack()
                                        /usr/local/go/src/runtime/debug/stack.go:26 +0x5e
                                github.com/stretchr/testify/assert.didPanic.func1.1()
                                        /home/ayush/go/pkg/mod/github.com/stretchr/testify@v1.7.0/assert/assertions.go:1013 +0x69
                                panic({0x775f20?, 0x2ce332042978?})
                                        /usr/local/go/src/runtime/panic.go:860 +0x13a
                                github.com/apache/fory/go/fory.(*MetaStringResolver).ReadMetaStringBytes(0x2ce33203c3c0, 0x2ce3320bec80, 0x2ce33203c420)
                                        /home/ayush/contributions/apache/fory/go/fory/meta_string_resolver.go:130 +0x827
                                github.com/apache/fory/go/fory.TestMetaStringResolver_NegativeIndexPanic.func1()
                                        /home/ayush/contributions/apache/fory/go/fory/security_test.go:58 +0x28
                                github.com/stretchr/testify/assert.didPanic.func1(0x2ce3321ddda0?, 0x5cb44e?, 0x41e07f?, 0x73ce60?)
                                        /home/ayush/go/pkg/mod/github.com/stretchr/testify@v1.7.0/assert/assertions.go:1018 +0x4c
                                github.com/stretchr/testify/assert.didPanic(0xa5a5c0?)
                                        /home/ayush/go/pkg/mod/github.com/stretchr/testify@v1.7.0/assert/assertions.go:1020 +0x36
                                github.com/stretchr/testify/assert.NotPanics({0x7b7e40, 0x2ce3321a0908}, 0x2ce3321ddd70, {0x2ce3320a5f48, 0x1, 0x1})
                                        /home/ayush/go/pkg/mod/github.com/stretchr/testify@v1.7.0/assert/assertions.go:1091 +0x7e
                                github.com/stretchr/testify/require.NotPanics({0x7b8b68, 0x2ce3321a0908}, 0x2ce3321ddd70, {0x2ce332094f48, 0x1, 0x1})
                                        /home/ayush/go/pkg/mod/github.com/stretchr/testify@v1.7.0/require/require.go:1456 +0xa6
                                github.com/apache/fory/go/fory.TestMetaStringResolver_NegativeIndexPanic(0x2ce3321a0908)
                                        /home/ayush/contributions/apache/fory/go/fory/security_test.go:57 +0x194
                                testing.tRunner(0x2ce3321a0908, 0x7b3580)
                                        /usr/local/go/src/testing/testing.go:2036 +0xea
                                created by testing.(*T).Run in goroutine 1
                                        /usr/local/go/src/testing/testing.go:2101 +0x4c5
                Test:           TestMetaStringResolver_NegativeIndexPanic
                Messages:       MetaStringResolver should not panic on negative index
--- FAIL: TestMetaStringResolver_NegativeIndexPanic (0.00s)
FAIL
exit status 1
FAIL    github.com/apache/fory/go/fory  0.004s
```

--- a/go/fory/meta_string_resolver.go
+++ b/go/fory/meta_string_resolver.go
@@ -124,7 +124,7 @@ func (r *MetaStringResolver) ReadMetaStringBytes(buf *ByteBuffer, ctxErr *Error)
 	length := int16(header >> 1)
 	if header&1 != 0 {
 		index := int(length) - 1
-		if index >= len(r.dynamicIDToEnumString) {
+		if index < 0 || index >= len(r.dynamicIDToEnumString) {
 			return nil, fmt.Errorf("invalid dynamic index: %d", index)
 		}
 		return r.dynamicIDToEnumString[index], nil

--- a/go/fory/meta_string_resolver_test.go
+++ b/go/fory/meta_string_resolver_test.go
@@ -1,0 +1,67 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package fory
+
+import (
+	"testing"
+	"github.com/stretchr/testify/require"
+)
+
+// TestMetaStringResolverNegativeIndexPanic reproduces the CRITICAL security bug 
+// in MetaStringResolver where a header of 1 triggers a -1 index panic.
+func TestMetaStringResolverNegativeIndexPanic(t *testing.T) {
+	resolver := NewMetaStringResolver()
+	buffer := NewByteBuffer(nil)
+	
+	// header = 1 means (header & 1 != 0) is true (it's a reference)
+	// and length = header >> 1 = 0.
+	// index = length - 1 = -1.
+	buffer.WriteVarUint32Small7(1)
+	buffer.SetReaderIndex(0)
+
+	var ctxErr Error
+	// This should NOT panic. The fix handles the negative index.
+	require.NotPanics(t, func() {
+		_, err := resolver.ReadMetaStringBytes(buffer, &ctxErr)
+		if err == nil {
+			t.Errorf("Expected error for negative index, got nil")
+		}
+	}, "MetaStringResolver should not panic on negative index")
+}
+
+// TestMetaStringResolverBoundaryRegression verifies that the smallest valid
+// dynamic index (resulting in index 0) still resolves correctly.
+func TestMetaStringResolverBoundaryRegression(t *testing.T) {
+	resolver := NewMetaStringResolver()
+	
+	// Add a string to the cache so len is 1
+	m := NewMetaStringBytes([]byte("test"), 123)
+	resolver.dynamicIDToEnumString = append(resolver.dynamicIDToEnumString, m)
+
+	// Craft a header that points to index 0
+	// header = (index + 1) << 1 | 1 
+	// For index 0: (0 + 1) << 1 | 1 = 3
+	buffer := NewByteBuffer(nil)
+	buffer.WriteVarUint32Small7(3) 
+	buffer.SetReaderIndex(0)
+
+	var ctxErr Error
+	result, err := resolver.ReadMetaStringBytes(buffer, &ctxErr)
+	require.NoError(t, err)
+	require.Equal(t, m, result, "Should correctly resolve the first dynamic string (index 0)")
+}

--- a/go/fory/meta_string_resolver_test.go
+++ b/go/fory/meta_string_resolver_test.go
@@ -18,16 +18,16 @@
 package fory
 
 import (
-	"testing"
 	"github.com/stretchr/testify/require"
+	"testing"
 )
 
-// TestMetaStringResolverNegativeIndexPanic reproduces the CRITICAL security bug 
+// TestMetaStringResolverNegativeIndexPanic reproduces the CRITICAL security bug
 // in MetaStringResolver where a header of 1 triggers a -1 index panic.
 func TestMetaStringResolverNegativeIndexPanic(t *testing.T) {
 	resolver := NewMetaStringResolver()
 	buffer := NewByteBuffer(nil)
-	
+
 	// header = 1 means (header & 1 != 0) is true (it's a reference)
 	// and length = header >> 1 = 0.
 	// index = length - 1 = -1.
@@ -48,16 +48,16 @@ func TestMetaStringResolverNegativeIndexPanic(t *testing.T) {
 // dynamic index (resulting in index 0) still resolves correctly.
 func TestMetaStringResolverBoundaryRegression(t *testing.T) {
 	resolver := NewMetaStringResolver()
-	
+
 	// Add a string to the cache so len is 1
 	m := NewMetaStringBytes([]byte("test"), 123)
 	resolver.dynamicIDToEnumString = append(resolver.dynamicIDToEnumString, m)
 
 	// Craft a header that points to index 0
-	// header = (index + 1) << 1 | 1 
+	// header = (index + 1) << 1 | 1
 	// For index 0: (0 + 1) << 1 | 1 = 3
 	buffer := NewByteBuffer(nil)
-	buffer.WriteVarUint32Small7(3) 
+	buffer.WriteVarUint32Small7(3)
 	buffer.SetReaderIndex(0)
 
 	var ctxErr Error

--- a/go/fory/ref_resolver.go
+++ b/go/fory/ref_resolver.go
@@ -277,6 +277,9 @@ func (r *RefResolver) GetReadObject(refId int32) reflect.Value {
 	if refId < 0 {
 		return r.readObject
 	}
+	if int(refId) >= len(r.readObjects) {
+		return reflect.Value{}
+	}
 	return r.readObjects[refId]
 }
 

--- a/go/fory/ref_resolver_test.go
+++ b/go/fory/ref_resolver_test.go
@@ -189,7 +189,7 @@ func TestRefTrackingLargeCount(t *testing.T) {
 func TestRefResolverOOBPanic(t *testing.T) {
 	resolver := newRefResolver(true)
 	buffer := NewByteBuffer(nil)
-	
+
 	// Craft a buffer with RefFlag (-2) followed by a huge RefID (9999)
 	buffer.WriteInt8(RefFlag)
 	buffer.WriteVarUint32(9999)
@@ -207,7 +207,7 @@ func TestRefResolverOOBPanic(t *testing.T) {
 func TestRefResolverBoundaryRegression(t *testing.T) {
 	resolver := newRefResolver(true)
 	val := reflect.ValueOf("test")
-	
+
 	// Fill the resolver with 2 objects
 	id0, _ := resolver.PreserveRefId()
 	resolver.SetReadObject(id0, val)

--- a/go/fory/ref_resolver_test.go
+++ b/go/fory/ref_resolver_test.go
@@ -183,3 +183,38 @@ func TestRefTrackingLargeCount(t *testing.T) {
 		})
 	}
 }
+
+// TestRefResolverOOBPanic reproduces the CRITICAL security bug in RefResolver
+// where a huge refId in the buffer triggers an OOB panic.
+func TestRefResolverOOBPanic(t *testing.T) {
+	resolver := newRefResolver(true)
+	buffer := NewByteBuffer(nil)
+	
+	// Craft a buffer with RefFlag (-2) followed by a huge RefID (9999)
+	buffer.WriteInt8(RefFlag)
+	buffer.WriteVarUint32(9999)
+	buffer.SetReaderIndex(0)
+
+	var ctxErr Error
+	// This should NOT panic. The fix handles the invalid index gracefully.
+	require.NotPanics(t, func() {
+		resolver.ReadRefOrNull(buffer, &ctxErr)
+	}, "RefResolver.GetReadObject should not panic on OOB index")
+}
+
+// TestRefResolverBoundaryRegression verifies that valid boundary indices
+// (0 and len-1) still resolve correctly after the security fix.
+func TestRefResolverBoundaryRegression(t *testing.T) {
+	resolver := newRefResolver(true)
+	val := reflect.ValueOf("test")
+	
+	// Fill the resolver with 2 objects
+	id0, _ := resolver.PreserveRefId()
+	resolver.SetReadObject(id0, val)
+	id1, _ := resolver.PreserveRefId()
+	resolver.SetReadObject(id1, val)
+
+	// Verify boundary indices still work
+	require.Equal(t, val.String(), resolver.GetReadObject(0).String(), "Should resolve index 0")
+	require.Equal(t, val.String(), resolver.GetReadObject(1).String(), "Should resolve index len-1")
+}

